apps/btc_app/get_public_key.c

```diff
@@ -0,0 +1,293 @@
+ * sends the uncompressed public key to the host. Prior to sending the public
+ * key, the function derives and verifies public address in the respective
+ * encoding format.
+ *
+ * @param query Reference to the query received from the host
+ * @param seed The wallet seed generated from X1 Card
+ *
+ * @return bool Indicating if the public key was sent successfully.
+ * @retval true The user confirmed his/her intention to export address
+ * @retval false The user either rejected the prompt or a P0 event occurred
+ */
+static bool confirm_and_send(const btc_query_t *query, const uint8_t *seed);
+
+/*****************************************************************************
+ * STATIC VARIABLES
+ *****************************************************************************/
+
+static bool check_which_request(const btc_query_t *query,
+                                pb_size_t which_request) {
+  if (which_request != query->get_public_key.which_request) {
+    btc_send_error(ERROR_COMMON_ERROR_CORRUPT_DATA_TAG,
+                   ERROR_DATA_FLOW_INVALID_REQUEST);
+    return false;
+  }
+
+  return true;
+}
+
+static bool validate_request_data(btc_get_public_key_request_t *request) {
+  bool status = true;
+
+  // TODO: Enable btc/coin specific check
+  if (false ==
+      verify_receive_derivation_path(request->initiate.derivation_path,
+                                     request->initiate.derivation_path_count)) {
+    btc_send_error(ERROR_COMMON_ERROR_CORRUPT_DATA_TAG,
+                   ERROR_DATA_FLOW_INVALID_DATA);
+    status = false;
+  }
+  return status;
+}
+
+static int btc_get_address(const uint8_t *seed,
+                           const uint32_t *path,
+                           uint32_t path_length,
+                           uint8_t *public_key,
+                           char *address) {
+  HDNode node = {0};
+  uint8_t address_version = 0;
+  uint32_t purpose_index = path[0];
+  uint32_t coin_index = path[1];
+  int address_length = 0;
+
+  if (false ==
+      derive_hdnode_from_path(path, path_length, SECP256K1_NAME, seed, &node)) {
+    // send unknown error; unknown failure reason
+    btc_send_error(ERROR_COMMON_ERROR_UNKNOWN_ERROR_TAG, 1);
+    memzero(&node, sizeof(HDNode));
+    return address_length;
+  }
+
+  switch (purpose_index) {
+    case NATIVE_SEGWIT:
+      address_length = get_segwit_address(
+          node.public_key, sizeof(node.public_key), coin_index, address);
+      break;
+    case NON_SEGWIT:
+      get_version(purpose_index, coin_index, &address_version, NULL);
+      hdnode_get_address(&node, address_version, address, 35);
+      break;
+    // TODO: add support for taproot and segwit
+    default:
+      address_length = 0;
+      break;
+  }
+
+  if (NULL != public_key) {
+    memcpy(public_key, node.public_key, ECDSA_PUB_KEY_SIZE);
+  }
+  memzero(&node, sizeof(HDNode));
+  return address_length;
+}
+
+static bool confirm_and_send(const btc_query_t *query, const uint8_t *seed) {
```

Comment:
Please consider breaking in to 2 functions - 1 for confirmation and 1 for creating the public keys